### ISSUE TITLE
response: signature validation in SAML assertion failed 

### ISSUE THREAD

**Issue by manuelwaltschek** [onelogin/java-saml#181]
Hello there,

I am new to SAML and therefore tried to find a simple testing framework / mock for an IdP, since I need to implement a service provider for my application while I don't know the IdP implementation.

I found https://capriza.github.io/samling/samling.html which is a serverless (as-in client side only) SAML IdP for the purpose of testing SAML integrations.

The following xml shows the metadata of the samling IdP:
```xml
<?xml version="1.0"?>
<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="https://capriza.github.io/samling/samling.html">
  <md:IDPSSODescriptor WantAuthnRequestsSigned="false" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
    <md:KeyDescriptor use="signing">
      <ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
        <ds:X509Data>
          <ds:X509Certificate>MIICpzCCAhACCQDuFX0Db5iljDANBgkqhkiG9w0BAQsFADCBlzELMAkGA1UEBhMCVVMxEzARBgNVBAgMCkNhbGlmb3JuaWExEjAQBgNVBAcMCVBhbG8gQWx0bzEQMA4GA1UECgwHU2FtbGluZzEPMA0GA1UECwwGU2FsaW5nMRQwEgYDVQQDDAtjYXByaXphLmNvbTEmMCQGCSqGSIb3DQEJARYXZW5naW5lZXJpbmdAY2Fwcml6YS5jb20wHhcNMTgwNTE1MTgxMTEwWhcNMjgwNTEyMTgxMTEwWjCBlzELMAkGA1UEBhMCVVMxEzARBgNVBAgMCkNhbGlmb3JuaWExEjAQBgNVBAcMCVBhbG8gQWx0bzEQMA4GA1UECgwHU2FtbGluZzEPMA0GA1UECwwGU2FsaW5nMRQwEgYDVQQDDAtjYXByaXphLmNvbTEmMCQGCSqGSIb3DQEJARYXZW5naW5lZXJpbmdAY2Fwcml6YS5jb20wgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAJEBNDJKH5nXr0hZKcSNIY1l4HeYLPBEKJLXyAnoFTdgGrvi40YyIx9lHh0LbDVWCgxJp21BmKll0CkgmeKidvGlr3FUwtETro44L+SgmjiJNbftvFxhNkgA26O2GDQuBoQwgSiagVadWXwJKkodH8tx4ojBPYK1pBO8fHf3wOnxAgMBAAEwDQYJKoZIhvcNAQELBQADgYEACIylhvh6T758hcZjAQJiV7rMRg+Omb68iJI4L9f0cyBcJENR+1LQNgUGyFDMm9Wm9o81CuIKBnfpEE2Jfcs76YVWRJy5xJ11GFKJJ5T0NEB7txbUQPoJOeNoE736lF5vYw6YKp8fJqPW0L2PLWe9qTn8hxpdnjo3k6r5gXyl8tk=</ds:X509Certificate>
        </ds:X509Data>
      </ds:KeyInfo>
    </md:KeyDescriptor>
    <md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress</md:NameIDFormat>
    <md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://capriza.github.io/samling/samling.html"/>
    <md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://capriza.github.io/samling/samling.html"/>
  </md:IDPSSODescriptor>
</md:EntityDescriptor>
```

When I deploy java-saml-tookit-jspsample-2.3.1-SNAPSHOT.war on my wildfly 10.1.Final installation and click the login-button on the index.jsp I am redirected to samling to create the response which is sent to acs.jsp on my SP. 

Example response:

```xml
<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" Version="2.0" ID="_fe5225b6dd62e9e68027307236054e5e8cc1ebe900" IssueInstant="2018-08-30T10:25:31.516Z" InResponseTo="ONELOGIN_17c9e98a-ae76-46e4-a164-c9c7bafcfd67">
	<saml:Issuer>https://capriza.github.io/samling/samling.html</saml:Issuer>
	<samlp:Status>
		<samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
	</samlp:Status>
	<saml:Assertion xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" Version="2.0" ID="_36b11961be946499e198977f4fb4cbdbc4e7a8b599" IssueInstant="2018-08-30T10:25:31.337Z">
		<saml:Issuer>https://capriza.github.io/samling/samling.html</saml:Issuer>
		<ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
			<ds:SignedInfo>
				<ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				<ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
				<ds:Reference URI="#_36b11961be946499e198977f4fb4cbdbc4e7a8b599">
					<ds:Transforms>
						<ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
						<ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
					</ds:Transforms>
					<ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
					<ds:DigestValue>77HBm7F2KA2lVD7w6gHbMJG8nd4aBdNLl+UCqSR3Qys=</ds:DigestValue>
				</ds:Reference>
			</ds:SignedInfo>
			<ds:SignatureValue>Abzw/TQm4FKVmSYMbqssEkNaCDpisYcfIouqqnd07qqPkIPD8ik8LXfdkvZVwxA7hXdGP/hrQwXInRqA0GW1lJ2uQanGHsvQQuqUEDVsR76TI7/TMunDqfd6JOA54fiLoHyJbTvzEjtrzopNUpw/Spt8X5bWKgoX/RphaN/XUPQ=</ds:SignatureValue>
			<ds:KeyInfo>
				<ds:X509Data>
					<ds:X509Certificate>MIIDCTCCAnKgAwIBAgIBATANBgkqhkiG9w0BAQUFADBvMRQwEgYDVQQDEwtjYXByaXphLmNvbTELMAkGA1UEBhMCVVMxETAPBgNVBAgTCFZpcmdpbmlhMRMwEQYDVQQHEwpCbGFja3NidXJnMRAwDgYDVQQKEwdTYW1saW5nMRAwDgYDVQQLEwdTYW1saW5nMB4XDTE4MDgyOTE0MTE1MVoXDTE5MDgyOTE0MTE1MVowbzEUMBIGA1UEAxMLY2Fwcml6YS5jb20xCzAJBgNVBAYTAlVTMREwDwYDVQQIEwhWaXJnaW5pYTETMBEGA1UEBxMKQmxhY2tzYnVyZzEQMA4GA1UEChMHU2FtbGluZzEQMA4GA1UECxMHU2FtbGluZzCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEApHfyvytCYR9VFm3jzYk1O7ziRxbzMTtCvyGjAc2ljvYn0tqT23HeutIKLcHwqKsbD20atU9sdHDEdIz8qPm6NrjXXt82sDBmy+ODFRk1Ym5FDqlu4rDoI7PJ44LRY+5msKKhrddUauLYMJ2wvVx5nZParHgDPkEmXkgRPfMNWFkCAwEAAaOBtDCBsTAMBgNVHRMEBTADAQH/MAsGA1UdDwQEAwIC9DA7BgNVHSUENDAyBggrBgEFBQcDAQYIKwYBBQUHAwIGCCsGAQUFBwMDBggrBgEFBQcDBAYIKwYBBQUHAwgwEQYJYIZIAYb4QgEBBAQDAgD3MCUGA1UdEQQeMByGGmh0dHA6Ly9jYXByaXphLmNvbS9zYW1saW5nMB0GA1UdDgQWBBQdwtHvzEpuCZ5vSc/D8R7N6rv1ujANBgkqhkiG9w0BAQUFAAOBgQAnaRPNFvsZAEl/gtYSQYhywKJDh/RG9TIjLlDOsf2c3KnamLBREoPpBtjV0s+R+e3/CND0MSYKvFwQyLxnlkw+fjEwhn36if0ROBA2yNmNvB607D2k7tg5Izyvq7Rsnq5r8MNQIM6LzwlvBLFVziJlG+EFuiqp1Lgh7Fhn0O2W3w==</ds:X509Certificate>
				</ds:X509Data>
			</ds:KeyInfo>
		</ds:Signature>
		<saml:Subject>
			<saml:NameID Format="urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified">prismaprisma</saml:NameID>
			<saml:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
				<saml:SubjectConfirmationData NotOnOrAfter="2018-08-30T10:35:31.337Z" Recipient="http://localhost:8080/java-saml-tookit-jspsample-2.3.1-SNAPSHOT/acs.jsp" InResponseTo="ONELOGIN_17c9e98a-ae76-46e4-a164-c9c7bafcfd67"/>
			</saml:SubjectConfirmation>
		</saml:Subject>
		<saml:Conditions NotBefore="2018-08-30T10:25:31.337Z" NotOnOrAfter="2018-08-30T10:35:31.337Z">
			<saml:AudienceRestriction>
				<saml:Audience>http://localhost:8080/java-saml-tookit-jspsample-2.3.1-SNAPSHOT/metadata.jsp</saml:Audience>
			</saml:AudienceRestriction>
		</saml:Conditions>
		<saml:AuthnStatement AuthnInstant="2018-08-30T10:25:31.337Z" SessionNotOnOrAfter="2018-08-30T10:35:31.328Z" SessionIndex="_samling_392801_89142647636">
			<saml:AuthnContext>
				<saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:Password</saml:AuthnContextClassRef>
			</saml:AuthnContext>
		</saml:AuthnStatement>
	</saml:Assertion>
</samlp:Response>
```

As you can see there is another certificate for the signing of the response assertion than in the metadata of the IdP (which I set as property of onelogin.saml2.idp.x509cert in onelogin.saml.properties file)

Now when I post the response to the SP the certlist to validate the response only contains the certificate from the metadata and not the one which is included in the response. This leads to com.onelogin.saml2.util.Util.validateSignNode(Node, X509Certificate, String, String) in com.onelogin.saml2.util.Util.validateSign(Document, List<X509Certificate>, String, String, String) returning "false" and therefore to !auth.isAuthenticated() in acs.jsp.

A possible solution would be to read the certificate from the response and add it to certlist in com.onelogin.saml2.authn.SamlResponse.isValid(String). But it would only make sense if this behavior (having another certificate for signing responses than in the metadata) isn't samling-specific. 

What do you think?

Edit: this is not a copy of https://github.com/onelogin/java-saml/issues/175

**Comment by pitbulk** [comment]
At https://capriza.github.io/samling/samling.html#showMetadata

is published the SAML metadata of samling which contains its x509 public certificate:

```
MIICpzCCAhACCQDuFX0Db5iljDANBgkqhkiG9w0BAQsFADCBlzELMAkGA1UEBhMCVVMxEzARBgNVBAgMCkNhbGlmb3JuaWExEjAQBgNVBAcMCVBhbG8gQWx0bzEQMA4GA1UECgwHU2FtbGluZzEPMA0GA1UECwwGU2FsaW5nMRQwEgYDVQQDDAtjYXByaXphLmNvbTEmMCQGCSqGSIb3DQEJARYXZW5naW5lZXJpbmdAY2Fwcml6YS5jb20wHhcNMTgwNTE1MTgxMTEwWhcNMjgwNTEyMTgxMTEwWjCBlzELMAkGA1UEBhMCVVMxEzARBgNVBAgMCkNhbGlmb3JuaWExEjAQBgNVBAcMCVBhbG8gQWx0bzEQMA4GA1UECgwHU2FtbGluZzEPMA0GA1UECwwGU2FsaW5nMRQwEgYDVQQDDAtjYXByaXphLmNvbTEmMCQGCSqGSIb3DQEJARYXZW5naW5lZXJpbmdAY2Fwcml6YS5jb20wgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAJEBNDJKH5nXr0hZKcSNIY1l4HeYLPBEKJLXyAnoFTdgGrvi40YyIx9lHh0LbDVWCgxJp21BmKll0CkgmeKidvGlr3FUwtETro44L+SgmjiJNbftvFxhNkgA26O2GDQuBoQwgSiagVadWXwJKkodH8tx4ojBPYK1pBO8fHf3wOnxAgMBAAEwDQYJKoZIhvcNAQELBQADgYEACIylhvh6T758hcZjAQJiV7rMRg+Omb68iJI4L9f0cyBcJENR+1LQNgUGyFDMm9Wm9o81CuIKBnfpEE2Jfcs76YVWRJy5xJ11GFKJJ5T0NEB7txbUQPoJOeNoE736lF5vYw6YKp8fJqPW0L2PLWe9qTn8hxpdnjo3k6r5gXyl8tk=
```

That the value you MUST set at onelogin.saml2.idp.x509cert
as explained on the [documentation](https://github.com/onelogin/java-saml#settings)


The SAMLResponse generated by the IdP should contain signature with that value otherwise, something is wrong at samling tool.

**Comment by manuelwaltschek** [comment]
Thank you @pitbulk for your reply.

That's what I did, but the response is signed with the certificate that is sent via the response. When I set the value of onelogin.saml2.idp.x509cert to the cert in the response everything worked as expected, so there must be something wrong at samling tool. 

Isn't it possible that some implementations use different keys for encryption and signing responses?
